File: src/year2017/day5/day5.rs
//! [Day 5: A Maze of Twisty Trampolines, All Alike](https://adventofcode.com/2017/day/5)

struct Puzzle {
    jumps: Vec<i32>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            jumps: data.lines().map(|s| s.parse().unwrap()).collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let mut jumps = self.jumps.clone();
        let length = i32::try_from(jumps.len()).unwrap();
        let mut offset = 0;
        let mut n = 0;

        while 0 <= offset && offset < length {
            let index = usize::try_from(offset).unwrap();

            let jump = jumps[index];
            jumps[index] += 1;

            offset += jump;
            n += 1;
        }

        n
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let mut jumps = self.jumps.clone();
        let length = i32::try_from(jumps.len()).unwrap();
        let mut offset = 0;
        let mut n = 0;

        while 0 <= offset && offset < length {
            let index = usize::try_from(offset).unwrap();

            let jump = jumps[index];
            if jump >= 3 {
                jumps[index] -= 1;
            } else {
                jumps[index] += 1;
            }

            offset += jump;
            n += 1;
        }

        n
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 5);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 10);
    }
}

File: src/year2017/day5/day5.py
#!/usr/bin/env python3
# [Day 5: A Maze of Twisty Trampolines, All Alike](https://adventofcode.com/2017/day/5)

import atexit
import sys
import time
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


maze = list(map(int, data.splitlines()))
i = 0
n = 0
while 0 <= i < len(maze):
    offset = maze[i]
    maze[i] += 1
    i += offset
    n += 1
print(n)

maze = list(map(int, data.splitlines()))
i = 0
n = 0
while 0 <= i < len(maze):
    offset = maze[i]
    if offset >= 3:
        maze[i] -= 1
    else:
        maze[i] += 1
    i += offset
    n += 1
print(n)
